dandi/dandiarchive.py

```diff
@@ -22,8 +16,200 @@
 lgr = get_logger()
 
 
+class ParsedDandiURL(ABC, BaseModel):
+    api_url: AnyHttpUrl  # Does not have a trailing slash
+    dandiset_id: str
+    version_id: Optional[str] = None
+
+    @validator("api_url")
+    def _validate_api_url(cls, v: AnyHttpUrl) -> AnyHttpUrl:
+        return parse_obj_as(AnyHttpUrl, v.rstrip("/"))
+
+    def get_client(self) -> DandiAPIClient:
+        return DandiAPIClient(self.api_url)
+
+    def get_dandiset(self, client: DandiAPIClient) -> dict:
+        return client.get_dandiset(self.dandiset_id, self.get_version_id(client))
+
+    def get_version_id(self, client: DandiAPIClient) -> str:
+        if self.version_id is None:
+            r = client.get(f"/dandisets/{self.dandiset_id}/")
+            published_version = r["most_recent_published_version"]
+            if published_version:
+                return published_version["version"]
+            else:
+                return r["draft_version"]["version"]
```

Comment:
```suggestion
            version = r["most_recent_published_version"] or r["draft_version"]
            return version["version"]
```
so we centralize a bit happen we need to later make it into `version.get("version")` etc